04-materials/04-debugging.md

```diff
@@ -1,111 +1,379 @@
 - Inspecting DOM elements created by your extension
-- Examining CSS styles and computed values
-- Modifying elements for quick testing
+- Check computed values (the final calculated values after all CSS is applied)
+- Temporarily modify HTML or CSS to test fixes
+
+:::{hint}
+There are times where you might right-click on an element expecting to see the browser with the "Inspect" option, but are instead met with a JupyterLab menu. There are a few cases where JupyterLab prevents this default browser behavior and displays it's own context menu, but you can still access that browser menu by: holding `Shift` while right-clicking, or using the dev tools "Select an element" picker (using the keyboard shortcut `Cmd + Shift + C` for macOS or `Ctrl + Shift + C` for Windows/Linux).
+:::
 
-## 4. Best Practices
+---
+
+## 4️⃣ Best Practices
+
+### 🎯 Debugging Workflow
 
-### Debugging Workflow
 - Start simple: `console.log` first
+  When something breaks, you don't need to automatically reach for advanced tools. First adding a few `console.log()` statements can help you: verify that your function is being called, confirm the data has the shape you expect, or discern whether or not you're reaching a line of code that's problematic. Sometimes you'll realize what the issue is quickly with just these simple verification steps.
 - Progress to browser DevTools
+  If console logs aren't enough, open the browser console and look for errors. Read the stack trace and click on the file links to jump to the section of code where the issue originates from.
+- Checkout the Network tab if applicable
+  Look through the Network tab if data is involved. Make sure your API calls are succeeding and retrieving the data that you expect.
 - Use browser debugger breakpoints for complex issues
+  For more complex issues involving state changes, timing issues, or logic problems, set breakpoints in the Sources tab and step through the code. Watch variables change, examine the call stack and understand the flow.
 - Keep development environment consistent
+  Is the extension installed and enabled? Did your latest changes build successfully?
+
+### 🛡️ Prevention is better
 
-### Prevention is better
-- TypeScript type checking
-- Linting and formatting
-- Unit testing critical functionality
-- Clear error messages in your code
+Of course, even better than debugging, would be to have working code that behaves as you expect. A few practices that can save you time debugging are:
 
-## 5. Hands-on Exercise
+- **Let TypeScript help you**: The red squiggles that you might see in your editor are important. Type errors are often caught at compile time and don't become an issue as you're running your extension code in the browser.
+- **Use a linter**: Tools like ESLint and Prettier catch common mistakes and help enforce consistent patterns. They'll give you warnings about unused variables or potential null reference errors.
```

Comment:
```suggestion
- **Use a linter**: Tools like ESLint catch common mistakes and help enforce consistent patterns. They'll give you warnings about unused variables or potential null reference errors.
```

In my mental model, prettier is a formatter and can't catch _bugs_, which is what I think of as a linter's job. I don't know if this distinction is important for learners at this stage though. What do you think? I do think if we don't make the distinction, we should just not mention prettier to avoid future confusion ;)

Reply:
Maybe we could define linting in the glossary?